apm-agent-plugin-sdk/src/main/java/co/elastic/apm/agent/sdk/internal/util/IOUtils.java

```diff
@@ -296,16 +305,23 @@ public interface ByteSourceReader<S> {
      * @return null, if the charset is not known/supported. Otherwise the result of the decoding operation.
      */
     @Nullable
-    public static CoderResult decode(ByteBuffer input, CharBuffer output, String charsetName) {
+    public static CoderResult decode(List<ByteBuffer> input, CharBuffer output, String charsetName) {
         try (ObjectHandle<CharsetDecoder> decoderHandle = getPooledCharsetDecoder(charsetName)) {
             if (decoderHandle == null) {
                 return null; //charset is unsupported
             }
             CharsetDecoder charsetDecoder = decoderHandle.get();
             try {
-                final CoderResult coderResult = charsetDecoder.decode(input, output, true);
-                charsetDecoder.flush(output);
-                return coderResult;
+                Iterator<ByteBuffer> it = input.iterator();
+                while (it.hasNext()) {
```

Comment:
[nit] I think we tend to rely more on traditional for loop to avoid allocating an iterator.